analyzers/src/SonarAnalyzer.Common/Rules/ThrowReservedExceptionsBase.cs

```diff
@@ -18,40 +18,37 @@
  * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
  */
 
-namespace SonarAnalyzer.Rules
+namespace SonarAnalyzer.Rules;
+
+public abstract class ThrowReservedExceptionsBase<TSyntaxKind> : SonarDiagnosticAnalyzer<TSyntaxKind>
+    where TSyntaxKind : struct
 {
-    public abstract class ThrowReservedExceptionsBase : SonarDiagnosticAnalyzer
-    {
-        internal const string DiagnosticId = "S112";
-        internal const string MessageFormat = "'{0}' should not be thrown by user code.";
+    protected const string DiagnosticId = "S112";
 
-        internal static readonly ImmutableArray<KnownType> ReservedExceptionTypeNames =
-            ImmutableArray.Create(
-                KnownType.System_Exception,
-                KnownType.System_ApplicationException,
-                KnownType.System_SystemException,
-                KnownType.System_ExecutionEngineException,
-                KnownType.System_IndexOutOfRangeException,
-                KnownType.System_NullReferenceException,
-                KnownType.System_OutOfMemoryException
-            );
+    private readonly ImmutableArray<KnownType> reservedExceptions = ImmutableArray.Create(
+        KnownType.System_Exception,
+        KnownType.System_ApplicationException,
+        KnownType.System_SystemException,
+        KnownType.System_ExecutionEngineException,
+        KnownType.System_IndexOutOfRangeException,
+        KnownType.System_NullReferenceException,
+        KnownType.System_OutOfMemoryException);
 
-        protected void ReportReservedExceptionCreation(SonarSyntaxNodeReportingContext context,
-            SyntaxNode throwStatementExpression)
-        {
-            if (throwStatementExpression == null ||
-                !IsObjectCreation(throwStatementExpression))
-            {
-                return;
-            }
+    protected abstract bool IsObjectCreation(SyntaxNode throwStatementExpression);
```

Comment:
I would just add a private method here instead of this (or even in-line it in the if statements, even better):
```cs
private bool IsObjectCreation(SyntaxNode throwStatementExpression) =>
    Language.Syntax.IsAnyKind(throwStatementExpression, Language.SyntaxKind.ObjectCreationExpressions);
```
The only difference is that it will check for ImplicitObjectCreationExpression as well for C# which might be superfluous for this rule, but it's more concise

Reply:
Nice